src/main/scala/firrtl/passes/InferReadWrite.scala

```diff
@@ -0,0 +1,182 @@
+     copyright notice, this list of conditions and the following
+     two paragraphs of disclaimer in the documentation and/or other materials
+     provided with the distribution.
+   * Neither the name of the Regents nor the names of its contributors
+     may be used to endorse or promote products derived from this
+     software without specific prior written permission.
+IN NO EVENT SHALL REGENTS BE LIABLE TO ANY PARTY FOR DIRECT, INDIRECT,
+SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES, INCLUDING LOST PROFITS,
+ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN IF
+REGENTS HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+REGENTS SPECIFICALLY DISCLAIMS ANY WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+A PARTICULAR PURPOSE. THE SOFTWARE AND ACCOMPANYING DOCUMENTATION, IF
+ANY, PROVIDED HEREUNDER IS PROVIDED "AS IS". REGENTS HAS NO OBLIGATION
+TO PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR
+MODIFICATIONS.
+*/
+
+package firrtl.passes
+
+import scala.collection.mutable.{ArrayBuffer, HashSet, HashMap}
+import com.typesafe.scalalogging.LazyLogging
+
+import firrtl._
+import firrtl.ir._
+import firrtl.Mappers._
+import firrtl.PrimOps._
+import Annotations._
+
+case class InferReadWriteAnnotation(t: String, tID: TransID = TransID(-1))
+    extends Annotation with Loose with Unstable {
+  val target = CircuitName(t)
+  def duplicate(n: Named) = this.copy(t=n.name)
+}
+
+object InferReadWritePass extends Pass {
+  def name = "Infer ReadWrite Ports"
+
+  def inferReadWrite(m: Module) = {
+    import WrappedExpression.we
+    val connects = HashMap[String, Expression]()
+    val repl = HashMap[String, Expression]()
+    val stmts = ArrayBuffer[Statement]()
+    val zero = we(UIntLiteral(0, IntWidth(1)))
+    val one = we(UIntLiteral(1, IntWidth(1)))
+
+    def analyze(s: Statement): Unit = s match {
+      case s: Connect  =>
+        connects(s.loc.serialize) = s.expr
+      case s: PartialConnect =>
+        connects(s.loc.serialize) = s.expr
+      case s: DefNode =>
+        connects(s.name) = s.value
+      case s: Block =>
+        s.stmts foreach analyze
+      case _ =>
+    }
+
+    def getProductTermsFromExp(e: Expression): Seq[Expression] =
+      e match {
+        // No ConstProp yet...
+        case Mux(cond, tval, fval, _) if we(tval) == one && we(fval) == zero =>
+          cond +: getProductTerms(cond.serialize)
+        case DoPrim(op, args, consts, tpe) if op == And =>
+          e +: (args flatMap getProductTermsFromExp)
+        case _: WRef | _: SubField | _: SubIndex | _: SubAccess =>
+          e +: getProductTerms(e.serialize)
+        case _ =>
+          List(e)
+      }
+
+    def getProductTerms(node: String): Seq[Expression] =
+      if (connects contains node) getProductTermsFromExp(connects(node)) else Nil
+
+    def checkComplement(a: Expression, b: Expression) = (a, b) match {
```

Comment:
Add comment explaining cases, something like:
a complement of not(a)
a complement of eq(a,0)